source/mesh_deformation/fastscape.cc

```diff
@@ -0,0 +1,1739 @@
+              // If using the center slice, find velocities from the row closest to the center.
+              if (center_slice)
+                {
+                  const int index = i+nx*(round((ny-1)/2));
+                  V2[i-1] = values[index];
+                }
+              // Here we use average velocities across the y nodes, excluding the ghost nodes (top and bottom row).
+              // Note: If ghost nodes are turned off, boundary effects may influence this.
+              else
+                {
+                  for (int ys=1; ys<(ny-1); ++ys)
+                    {
+                      const int index = i+nx*ys;
+                      V2[i-1] += values[index];
+                    }
+                  V2[i-1] = V2[i-1]/(ny-2);
+                }
+            }
+
+          for (unsigned int i=0; i<data_table.size()[1]; ++i)
+            {
+              idx[1] = i;
+
+              for (unsigned int j=0; j<(data_table.size()[0]); ++j)
+                {
+                  idx[0] = j;
+
+                  // Convert back to m/s.
+                  data_table(idx) = V2[j] / year_in_seconds;
+                }
+            }
+        }
+      else
+        {
+          // Indexes through z, y, and then x.
+          for (unsigned int k=0; k<data_table.size()[2]; ++k)
+            {
+              idx[2] = k;
+
+              for (unsigned int i=0; i<data_table.size()[1]; ++i)
+                {
+                  idx[1] = i;
+
+                  for (unsigned int j=0; j<data_table.size()[0]; ++j)
+                    {
+                      idx[0] = j;
+
+                      // Convert back to m/s.
+                      data_table(idx) = values[(nx+1)*use_ghost_nodes+nx*i+j] / year_in_seconds;
+
+                    }
+                }
+            }
+        }
+
+      return data_table;
+    }
+
+
+
+    template <int dim>
+    void FastScape<dim>::read_restart_files(std::vector<double> &h,
+                                            std::vector<double> &b,
+                                            std::vector<double> &sf) const
+    {
+      this->get_pcout() << "   Loading FastScape restart file... " << std::endl;
+
+      // Create variables for output directory and restart file
+      std::string dirname = this->get_output_directory();
+      const std::string restart_filename = dirname + "fastscape_h_restart.txt";
+      const std::string restart_filename_basement = dirname + "fastscape_b_restart.txt";
+      const std::string restart_filename_silt_fraction = dirname + "fastscape_sf_restart.txt";
+      const std::string restart_filename_time = dirname + "fastscape_last_output_time.txt";
+
+      // Load in h values.
+      std::ifstream in;
+      in.open(restart_filename.c_str());
+      if (in)
+        {
+          unsigned int line = 0;
+
+          while (line < array_size)
+            {
+              in >> h[line];
+              line++;
+            }
+
+          in.close();
+        }
+
```

Comment:
What do you do if `in` can not be read from? I think you want
```
  AssertThrow (in, ExcIO());
```
before this block, and then you can skip the `if (in)`.